url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/optimizer/dataflow/DataFlowAnalyzer.java

```diff
@@ -0,0 +1,321 @@
+                    assignedColors.put( edge , currentColor );
+                } 
+            }
+        }
+        return currentColor <= colorCount;
+    }
+
+    public void assignLiveVariables(ControlFlowGraph graph) 
+    {
+        List<WorkListEntry> workList = new ArrayList<>();
+        Set<AliveVar> inState = new HashSet<>();
+        
+        for ( BasicBlock b : graph.getFunctionExit().getPredecessors() ) {
+            System.out.println("Adding to work queue: "+b);
+            workList.add( new WorkListEntry(b,inState ) );
+        }
+
+        final Set<BasicBlock> visited = new HashSet<>();
+        while ( ! workList.isEmpty() ) 
+        {
+            WorkListEntry entry = workList.remove(0);
+            processWorkListItem( visited , entry , workList );
+        }
+    }
+    
+    public static Set<AliveVar> getAliveVars(BasicBlock block) 
+    {
+        Set<AliveVar>  result = (Set<AliveVar>) block.getMetaData().get(ALIVE_VARS_METADATA_KEY);
+        return result == null ? Collections.<AliveVar>emptySet() : result;
+    }
+    
+    private Set<AliveVar> processWorkListItem(Set<BasicBlock> visited, WorkListEntry entry, List<WorkListEntry> workList) 
+    {
+        Set<AliveVar> newOutState  = getReadVars( entry.block );
+        Set<AliveVar> writtenVars = getWrittenVars( entry.block );
+        
+        entry.block.getMetaData().put(ALIVE_VARS_METADATA_KEY,newOutState);
+        
+        if ( ! newOutState.equals( entry.inState ) || ( newOutState.isEmpty() && writtenVars.isEmpty() ) ) 
+        {
+            for ( BasicBlock b : entry.block.getPredecessors() ) 
+            {
+                if ( ! visited.contains( b ) ) 
+                {
+                    System.out.println("Adding to work queue: "+b);
+                    visited.add( b );
+                    workList.add( new WorkListEntry( b , newOutState ) );
+                }
+            }
+        }
+        return newOutState;
+    }
+    
+    private Set<AliveVar> getReadVars(BasicBlock block) 
+    {
+        Set<AliveVar> readVars = new HashSet<>();
+        for ( ASTNode child : block.getASTNodes() ) 
+        {
+            if ( child instanceof VariableReferenceNode ) 
+            {
+                final VariableReferenceNode ref = (VariableReferenceNode ) child;
+                // check if we're on the LHS of an assignment
+                ASTNode previous=ref;
+                ASTNode current = ref.getParent();
+                while( current != null && ! isAssignmentOperator( current ) ) 
+                {
+                    previous = current;
+                    current = current.getParent();
+                }
+                final boolean onLHS = current != null && isAssignmentOperator( current ) && current.indexOf( previous ) == 0;
+                if ( ! onLHS ) {
+                    readVars.add( new AliveVar(ref.getDefinitionSite() ) );
+                }
+            }
+        }
+        return readVars;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 274:294, 299:319.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.